### PR TITLE
[TACHYON-731] Use try lock in worker

### PR DESCRIPTION
**Pull request by calvinjia**
https://tachyon.atlassian.net/browse/TACHYON-731

### CODE REVIEW

**Comment by carsonwang** [comment]
Thanks for the quick fix, @calvinjia . The try lock solution looks good on the worker side. Can we also make any improvement on the client side? Most of the methods are synchronized in TachyonFS. A client like Spark usually uses a single TachyonFS object in several threads. As we saw in recent two dead lock situations, when one client thread was requesting space, other threads were not able to unlock a block as they were waiting for the TachyonFS lock. Is it possible to make the client be able to unlock a block while another thread is requesting space using the same TachyonFS object?

**Comment by gpang** [comment]
Jenkins, test this please.

**Comment by carsonwang** [comment]
@calvinjia , the dead lock issue doesn't happen again after applying this patch. Sometimes I did notice the error messages this printed out like "Failed to lockBlock..." and "Failed master delete block cmd...". When this occurred, the heartbeat thread blocked for several seconds. This also blocked other related threads for several seconds as we saw in the dead lock situation. The heartbeat time can take up to 20 seconds in some cases. What is the consideration for making the lock timeout 5 seconds? Does 1 second cause too many lock failures?

File: servers/src/main/java/tachyon/worker/block/BlockLockManager.java [hunk 1]
```diff
@@ -95,7 +98,21 @@ public long lockBlock(long userId, long blockId, BlockLockType blockLockType) th
...
+    boolean success;
+    try {
+      success = lock.tryLock(LOCK_ACQUIRE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException ie) {
+      // The UserLock implementation does not throw this exception, something is wrong if it happens
+      LOG.error("Interrupted exception in tryLock, this should not occur!");
+      throw new IOException(ie.getMessage(), ie.getCause());
+    }
+    if (!success) {
+      String errMsg = "Failed to lockBlock: " + blockId + " for user: " + userId + " in 5ms.";
```
Comment (carsonwang) [inline servers/src/main/java/tachyon/worker/block/BlockLockManager.java]: minor:  "5ms" -> "5s" or "5000 ms"

**Comment by calvinjia** [comment]
@carsonwang 
Thanks for the update. I use 5 seconds because I did not want to cause errors due to the block being locked for a move or something like that. It should be very rare for the block to be consistently read when there is a pending request to delete the block. What kind of workload are you running which deletes files often when they are still being accessed?

File: servers/src/main/java/tachyon/worker/block/BlockMasterSync.java [hunk 2]
```diff
@@ -194,7 +194,11 @@ private void handleMasterCommand(Command cmd) throws IOException {
       // Master requests blocks to be removed from Tachyon managed space.
       case Free:
         for (long block : cmd.mData) {
-          mBlockDataManager.removeBlock(Users.MASTER_COMMAND_USER_ID, block);
+          try {
+            mBlockDataManager.removeBlock(Users.MASTER_COMMAND_USER_ID, block);
+          } catch (IOException ioe) {
+            LOG.warn("Failed master delete block cmd for: " + block + " due to concurrent read.");
```
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/BlockMasterSync.java]: "delete" sounds like the block is being deleted forever. Maybe this should be called free or remove or something?

File: servers/src/main/java/tachyon/worker/block/BlockLockManager.java [hunk 3]
```diff
@@ -95,7 +98,21 @@ public long lockBlock(long userId, long blockId, BlockLockType blockLockType) th
     } else { // blockLockType == BlockLockType.WRITE
       lock = blockLock.writeLock();
     }
-    lock.lock();
+
+    // The block lock may be busy, wait up to five seconds to obtain it.
+    boolean success;
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockLockManager.java]: could you also update the javadoc of this method accordingly?

File: servers/src/main/java/tachyon/worker/block/BlockMasterSync.java (hunk 2 above)
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockMasterSync.java]: do we retry in this case?

File: servers/src/main/java/tachyon/worker/block/BlockLockManager.java (hunk 3 above)
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/BlockLockManager.java]: Updated, thanks.

File: servers/src/main/java/tachyon/worker/block/BlockMasterSync.java (hunk 2 above)
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/BlockMasterSync.java]: Updated, thanks. The master will request to delete the block in the next heartbeat.

File: servers/src/main/java/tachyon/worker/block/BlockMasterSync.java [hunk 4]
```diff
@@ -194,7 +194,11 @@ private void handleMasterCommand(Command cmd) throws IOException {
       // Master requests blocks to be removed from Tachyon managed space.
       case Free:
         for (long block : cmd.mData) {
-          mBlockDataManager.removeBlock(Users.MASTER_COMMAND_USER_ID, block);
+          try {
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockMasterSync.java]: Place `try` outside the for-loop and be more responsive in heartbeating?

Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockMasterSync.java]: you can discard this comment.

**Comment by apc999** [comment]
+1 LGTM

**Comment by carsonwang** [comment]
I was running a Spark workload which is the same one that mingfei used for performance testing. I think this is sufficient as a fix to 0.7.
+1

**Comment by gpang** [comment]
LGTM

**Pull request merged**